pyatv/protocols/companion/api.py

```diff
@@ -283,6 +291,90 @@ async def hid_command(self, down: bool, command: HidCommand) -> None:
+        y = min(y, int(TOUCHPAD_HEIGHT))
+        await self._send_event(
+            identifier="_hidT",
+            content={
+                "_ns": (time.time_ns() - self._base_timestamp),
+                "_tFg": 1,
+                "_cx": x,
+                "_tPh": mode.value,
+                "_cy": y,
+            },
+        )
+
+    async def touch_swipe(
+        self, start_x: int, start_y: int, end_x: int, end_y: int, duration_ms: int
+    ):
+        """Generate a touch swipe.
+
+         From start to end x,y coordinates (in range [0,1000])
+         in a given time (in milliseconds).
+
+        :param start_x: Start x coordinate
+        :param start_y: Start y coordinate
+        :param end_x: End x coordinate
+        :param end_y: Endi x coordinate
+        :param duration_ms: Time in milliseconds to reach the end coordinates
+        """
+        end_time = time.time_ns() + duration_ms * 1000000
+        x: float = float(start_x)
+        y: float = float(start_y)
+        await self.hid_event(int(x), int(y), TouchAction.Press)
+        sleep_time = float(TOUCHPAD_DELAY_MS / 1000)
+        current_time = time.time_ns()
+        while current_time < end_time:
+            x = x + float(end_x - x) * TOUCHPAD_DELAY_MS * 1000000 / (
+                end_time - current_time
+            )
+            y = y + float(end_y - y) * TOUCHPAD_DELAY_MS * 1000000 / (
+                end_time - current_time
+            )
+            x = max(x, 0)
+            y = max(y, 0)
+            x = min(x, TOUCHPAD_WIDTH)
+            y = min(y, TOUCHPAD_HEIGHT)
+            await self.hid_event(int(x), int(y), TouchAction.Hold)
+            await asyncio.sleep(sleep_time)
+            current_time = time.time_ns()
+        await self.hid_event(end_x, end_y, TouchAction.Release)
+
+    async def touch_action(self, x: int, y: int, mode: TouchAction):
+        """Generate a touch event to x,y coordinates (in range [0,1000]).
+
+        :param x: x coordinate
+        :param y: y coordinate
+        :param mode: touch mode (1: press, 3: hold, 4: release)
+        """
+        await self.hid_event(x, y, mode)
+
+    async def touch_click(self, action: InputAction):
+        """Send a touch click.
+
+        :param action: action mode single tap (0), double tap (1), or hold (2)
+        """
+        if action in [InputAction.SingleTap, InputAction.DoubleTap]:
+            count = 1 if action == InputAction.SingleTap else 2
+            for i in range(count):
+                await self._send_command("_hidC", {"_hBtS": 1, "_hidC": 6})
+                await asyncio.sleep(0.02)
+                await self._send_command("_hidC", {"_hBtS": 2, "_hidC": 6})
+                await self.hid_event(
+                    int(TOUCHPAD_WIDTH), int(TOUCHPAD_HEIGHT), TouchAction.Click
+                )
+        else:  # Hold
+            await self._send_command("_hidC", {"_hBtS": 1, "_hidC": 6})
+            await asyncio.sleep(1)
```

Comment:
Same question here, since this has another length than above (should they be the same?).

Reply:
Here this is hold command, the hold duration can vary but this is the minimal duration so that it will be taken into account